### DISCUSSION TITLE
Python Client + Node Server not working

### DISCUSSION THREAD

**Discussion by neeyanthkvk** [miguelgrinberg/python-socketio#1243] [Q&A] [answered]
I am trying to create a websocket framework where there is a server running in node connected to a python client. Examples shown below:

**client**

```
MARKETDATA_ADDRESS = 'http://localhost:3000'

async def main2():
    socket = socketio.AsyncClient(logger=True, engineio_logger=True)
    await socket.connect(MARKETDATA_ADDRESS)
    await socket.emit('new_exchange', {'id': 'hi'})
    await socket.emit('end')
    await socket.disconnect()

if __name__ == "__main__":
    asyncio.run(main2())
```

**server**

```
const server = require('http').createServer();
const io = require('socket.io')(server);

let exchanges = {};

function handle_connect(data) { 
  console.log('new connection');
  console.log(data);
  console.log('finished');
}

function handle_marketdata_update(data) { 
  console.log('received md update from exchange');
  console.log(data);
  console.log('finished');
}


io.on('connection', client => {
  console.log('hi');
  client.on('event', handle_connect);
  client.on('new_exchange',  data => {
    console.log(data);
    console.log(client);
  });
  client.on('update_md', handle_marketdata_update);
});

console.log('starting websocket server at port 3000');
server.listen(3000);
```

When running the client, the connection is established (as in, the 'hi' is logged), but the output of the message from 'new_exchange' is not given. 

Equally confusing is when I use `socket-io.client` (the node client framework)

```
const io = require("socket.io-client");
const socket = io("http://localhost:3000");

socket.emit('event', {'id': 'hi'})
console.log('done');
socket.disconnect();
```

The original 'hi' is not logged. I'm not sure where the inconsistency is coming from and what I am missing. Thank you!

**Comment by neeyanthkvk** [comment]
For context, when running both the server and python client in debug mode, I get the following statements:

**from the server, using `DEBUG=socket.io* node server.js`**

```
socket.io:server initializing namespace / +0ms
  socket.io:server creating engine.io instance with opts {"cleanupEmptyChildNamespaces":false,"path":"/socket.io"} +0ms
  socket.io:server attaching client serving req handler +1ms
starting websocket server at port 3000
  socket.io:server incoming connection with id 0cR7tHgQPEubYOuOAAAA +2s
  socket.io-parser decoded 0{} as {"type":0,"nsp":"/","data":{}} +0ms
  socket.io:client connecting to namespace / +0ms
  socket.io:namespace adding socket to nsp / +0ms
  socket.io:socket socket connected - writing packet +0ms
  socket.io:socket join room lOprCiFk2JSF7inXAAAB +0ms
  socket.io-parser encoding packet {"type":0,"data":{"sid":"lOprCiFk2JSF7inXAAAB"},"nsp":"/"} +1ms
  socket.io-parser encoded {"type":0,"data":{"sid":"lOprCiFk2JSF7inXAAAB"},"nsp":"/"} as 0{"sid":"lOprCiFk2JSF7inXAAAB"} +0ms
hi
  socket.io:client client close with reason transport close +3ms
  socket.io:socket closing socket - reason transport close +2ms
```

**from the client, using `python lib.py`**
```
Attempting polling connection to http://localhost:3000/socket.io/?transport=polling&EIO=4
Polling connection accepted with {'sid': '0cR7tHgQPEubYOuOAAAA', 'upgrades': ['websocket'], 'pingInterval': 25000, 'pingTimeout': 20000, 'maxPayload': 1000000}
Engine.IO connection established
Sending packet MESSAGE data 0{}
Attempting WebSocket upgrade to ws://localhost:3000/socket.io/?transport=websocket&EIO=4
WebSocket upgrade was successful
Received packet MESSAGE data 0{"sid":"lOprCiFk2JSF7inXAAAB"}
Namespace / is connected
Emitting event "new_exchange" [/]
Sending packet MESSAGE data 2["new_exchange",{"id":"hi"}]
Emitting event "end" [/]
Sending packet MESSAGE data 2["end"]
Sending packet MESSAGE data 1
Sending packet CLOSE data None
Engine.IO connection dropped
Write loop: WebSocket connection was closed, aborting
Exiting write loop task
Server sent unexpected packet WSMsgType.CLOSING data None, aborting
Waiting for write loop task to end
Exiting read loop task
```

**Comment by miguelgrinberg** [comment, marked as answer]
The `emit` method in the Python client does not wait for the data to be transmitted. You are doing your emits and then immediately disconnecting, without giving time to the messages to arrive.

You have a few options to address this:

1. Add a sleep after the emits and before the disconnect. This is technically a hack, since you have no way to know how long it will take for the messages to be delivered.
2. Use acknowledgement callbacks to let the server notify you when the message was received. Callbacks in Python are strange and far from Pythonic, so this isn't a great option. If you go with this, the JS side also needs to be edited to use callbacks.
3. Use the `call()` method instead of `emit()`. The `call()` method encapsulates an emit with a callback, hiding all the ugliness behind a Pythonic interface. For this the JS side also needs to manage callbacks.

**Reply by neeyanthkvk** [reply]
This makes sense. The `call()` function is sending the message properly. What do you mean by callback? It is now stalling on the first message I send, and I assume that is because the client is waiting for a response from the server. How should I be managing these callbacks? Adding simple return statements to the handlers does not seem to solve the problem.

**Reply by neeyanthkvk** [reply]
Nevermind, I found the reference for how to do this. Thank you!
